### PR TITLE
infra: upgrade to rust-toolchain.toml

### PR DESCRIPTION
**Pull request by ppannuto**
### Pull Request Overview

Rustup configuration is now capable of installing desired components and platform targets for cross-compilation. Let's use the native stuff instead of homegrown scripting.

This is the followup PR promised in #3591.

### Testing Strategy

You're looking at it.

### TODO or Help Wanted

This pull request still needs to make sure CI works still :) 

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

### CODE REVIEW

File: tools/update_rust_version.sh [hunk 1]
```diff
@@ -30,12 +30,10 @@ NIGHTLY=nightly-$BEST_DATE
...
-# Set the Rust version in rust-toolchain file.
-echo $NIGHTLY > rust-toolchain
-
 # Update all relevant files with the new version string.
 # Note, x-platform `sed -i` has odd, but particular syntax
 # https://stackoverflow.com/questions/5694228/sed-in-place-flag-that-works-both-on-mac-bsd-and-linux
+sed -i._SED_HACK "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" rust-toolchain.toml
 sed -i._SED_HACK "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" .vscode/settings.json
 sed -i._SED_HACK "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" doc/Getting_Started.md
 sed -i._SED_HACK "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" rust-toolchain
```
Comment (bradjc) [inline tools/update_rust_version.sh:39]:
```suggestion
```

**Review by alistair23 — APPROVED** [review]

**Review by jrvanwhy — COMMENTED** [review]
You can also delete the `/rust-toolchain` line from `.lcignore`.

**Review by jrvanwhy — APPROVED** [review]

**Comment by lschuermann** [comment]
I'll need to update the `shell.nix` to switch from `rust-toolchain` to `rust-toolchain.toml`, marking as blocked.

File: tools/update_rust_version.sh (hunk 1 above)
Comment (ppannuto) [reply tools/update_rust_version.sh]: Updated commit

File: rust-toolchain.toml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Licensed under the Apache License, Version 2.0 or the MIT License.
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+# Copyright Tock Contributors 2022.
```
Comment (jrvanwhy) [inline rust-toolchain.toml:3]: nit: 2023

**Review by jrvanwhy — APPROVED** [review]

File: rust-toolchain.toml (hunk 2 above)
Comment (ppannuto) [reply rust-toolchain.toml]: copy/paste fail, thanks/fixed

**Review by bradjc — DISMISSED** [review]

**Review by alevy — DISMISSED** [review]

**Comment by lschuermann** [comment]
`shell.nix` now reads the `rust-toolchain.toml`, and I‘ve rebased this on the update to the 2023-07-30 toolchain.

**Review by lschuermann — APPROVED** [review]

**Review by lschuermann — APPROVED** [review]

**Review by lschuermann — DISMISSED** [review]

**Review by lschuermann — APPROVED** [review]

**Review by twilfredo — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

**Pull request merged**